syntax = "proto3";

package agent_manager_grpc;

enum ModelProvider {
    ANTHROPIC = 0;
    BEDROCK = 1;
}
message Cookie {
    map<string, string> cookie_data = 1;
}

message RunAgentRequest {
    string prompt = 1;
    string session_id = 2;
    // control flag to distinguish frontend requests
    bool is_chat_request = 3;
    optional string request_api_key = 4;
    optional string parent_span_context = 5;
    optional ModelProvider model_provider = 6;
    optional string model = 7;
    optional bool enable_thinking = 8;
    optional bool return_screenshots = 9;
    repeated Cookie cookies = 10;
}

message ActionResult {
    optional bool is_done = 1;
    optional string content = 2;
    optional string error = 3;
    optional bool give_control = 4;
}

message StepChunkContent {
    ActionResult action_result = 1;
    string summary = 2;
    string trace_id = 3;
    optional string screenshot = 4;
}

message AgentOutput {
    ActionResult result = 1;
    repeated Cookie cookies = 2;
    optional string trace_id = 3;
    optional uint64 step_count = 4;
}

message RunAgentResponseStreamChunk {
    oneof chunk_type {
        StepChunkContent step_chunk_content = 1;
        AgentOutput agent_output = 2;
    }
}

service AgentManagerService {
    rpc RunAgent(RunAgentRequest) returns (AgentOutput);
    rpc RunAgentStream(RunAgentRequest) returns (stream RunAgentResponseStreamChunk);
}
